Automattic/jetpack
# Sort Arrays by keys before generating callable checksums

The checksum we generate are different based on the order of Associated Arrays. This update applies a ksort on arrays to ensure we have consistent order.

Fixes # https://wp.me/p1HpG7-9UI

#### Changes proposed in this Pull Request:

* Applies a ksort when generating checksums to ensure duplicate updates are not sent for callables.

#### Does this pull request change what data or activity we track or use?


NO

#### Testing instructions:






* Checkout this branch
* Perform an update of a post (triggers sync)
* Verify that no callables where sent in Sync (Activity Log)
* Update a plugin (perform a change that modifies get_plugins callable function)
* Verify that the callable is sent via Sync
* Verify that in Activity Log we see a plugin__updated entry for the plugin

Perform updates to other callables (delete plugin, home/site_url, etc) and confirm they are sent via Sync.

#### Proposed changelog entry for your changes:

* Sync Bug Fix : Remove duplicate Activity Log entries by applying ksort during checksum generation.


#### Additional Notes

Please be aware that this change impacts how we determine if data should be sent to WP.com. It needs additional testing/scrutiny to ensure that no duplicate activity is generated for a site. 

Also from a development perspective is ksort the most effective approach. @bisko suggested the below which would be faster although I'm not sure if it would lead to additional or less collisions with the checksum.

```
$data = serialize( $values );
$data = explode('', $data);
$data = sort($data);
$checksum = md5($data);
```
